server/src/main/java/org/opensearch/action/support/TransportAction.java

```diff
@@ -88,31 +89,39 @@ public final Task execute(Request request, ActionListener<Response> listener) {
          */
         final Releasable unregisterChildNode = registerChildNode(request.getParentTask());
         final Task task;
+
         try {
             task = taskManager.register("transport", actionName, request);
         } catch (TaskCancelledException e) {
             unregisterChildNode.close();
             throw e;
         }
-        execute(task, request, new ActionListener<Response>() {
-            @Override
-            public void onResponse(Response response) {
-                try {
-                    Releasables.close(unregisterChildNode, () -> taskManager.unregister(task));
-                } finally {
-                    listener.onResponse(response);
+
+        ThreadContext.StoredContext storedContext = taskManager.taskExecutionStarted(task);
+        try {
+            execute(task, request, new ActionListener<Response>() {
```

Comment:
The second issue which strikes me: why task manager has register / unregister + cancellation but does not have an `execute(task, runnable)` method? If we feel this gap, the  `TransportAction` could just delegate the execution to `TaskManager` and not worrying about context manipulation - i will be sealed inside `TaskManager` only, thoughts?

Reply:
I was wondering something similar as all these pieces are spread across and execution doesn't seem to have common pipeline through which it flows through. 

But wouldn't want to include the change in this PR as I believe this part still needs more discussion and brain storming and might delay the current PR if we extend the scope.